Allow SparkSession to reuse SparkContext in the tests

avflor opened the pull request:
Fixes issue #1071 and #1088

Commit:
Allow SparkSession to reuse SparkContext in the tests

avflor commented:
@smurching Could you please take a look at this PR?

avflor committed:
Cleanup tracking environment after spark tests are run

smurching commented on tests/pyfunc/test_spark.py:
```diff
@@ -22,9 +22,11 @@
-    spark = pyspark.sql.SparkSession.builder \
-        .config(key="spark.python.worker.reuse", value=True) \
-        .master("local-cluster[2, 1, 1024]") \
+    spark = pyspark.sql.SparkSession.builder\
```
Ah good catch - I'm realizing that we intentionally decided to run tests for different model frameworks in different `pytest` invocations in https://github.com/mlflow/mlflow/pull/325/files (IIRC because the memory overhead from importing multiple different model frameworks in the same process could cause Travis tests to OOM). I was going to suggest that we start/stop SparkSessions in each test that uses them, but it seems starting multiple SparkSessions within the same Python process is not well-supported ([link](https://stackoverflow.com/questions/41491972/how-can-i-tear-down-a-sparksession-and-create-a-new-one-within-one-application)).

Thus maybe the right course of action here is to update our contributor guide (in CONTRIBUTING.rst) with test steps that more closely reflect our .travis.yml, as opposed to installing unused mleap dependencies in test_spark.py. I'll file a separate GitHub issue for updating CONTRIBUTING.rst - in this PR I'd suggest we remove changes to test_spark.py

smurching reviewed (commented):
Hi @avflor thanks for the PR & sorry for the delay! Had two small comments - mainly, I realized we'd actually intentionally decided to make Travis CI run model framework tests in separate Python processes, and so we should update our contributor documentation to reflect this, rather than try to make it possible to run all model framework tests within a single `pytest` invocation. 

I've filed https://github.com/mlflow/mlflow/issues/1096 to address this - let me know if you have any questions or thoughts, I'll aim to have a PR out for updating the contributor guide today. Thanks again for finding these issues - sorry you had to deal with them, and we really appreciate your help fixing them :)

avflor committed:
End run in mleap_model_log test

avflor committed:
Remove mleap jars from SparkSession in test_spark.py

avflor commented:
@smurching thank you very much for the comments and detailed explanations. I agree that updating the instructions in the contributor guide would definitely help the onboarding process and I'd be happy to verify that the new instructions work on my environment. 

I addressed your comments and updated the PR.

smurching reviewed (approved):
LGTM, merging to master, thanks @avflor!
